File: src/main/java/la/jitco/jana/FunctionGenerator.java
package la.jitco.jana;

public class FunctionGenerator extends Module {
    
    FunctionGenerator(int id)
    {
        super(id);
        type = ModuleType.FUNCTION_GENERATOR;
        name = "FunctionGenerator " + id;
        paramArray = new Parameter[] {
                         new NonnegativeFloatingPointParameter("baseFrequency"),
                         new OptionParameter("function",
                             new String[] {"Sine", "Saw", "Square"}),
                         new NonnegativeFloatingPointParameter("initPhase"),
                         new NonnegativeFloatingPointParameter("initLevel")};
        modInArray = new ModIn[] {
                         new ModIn(this, "frequency"),
                         new ModIn(this, "phase"),
                         new ModIn(this, "level")};
        modOutArray = new ModOut[] {
                         new ModOut(this, "mainOut"),
                         new ModOut(this, "auxOut1"),
                         new ModOut(this, "auxOut2"),
                         new ModOut(this, "auxOut3")};
        createMaps();
        weightTaskMultiplier = 10; // TODO: find actual value
        setWeight();
        
        paramArray[3].setValue("1.0");
        }

}

File: JobProcessing/FunctionGenerator.cpp
#include "FunctionGenerator.h"

#include <string>
#include "ModIn.h"
#include "ModOut.h"

Parameter& FunctionGenerator::getParameter(std::string parameterName)
{
    if (parameterName == "baseFrequency")
        return baseFrequency;
    else if (parameterName == "function")
        return function;
    else if (parameterName == "initPhase")
        return initPhase;
    else if (parameterName == "initLevel")
        return initLevel;
    else
        return baseFrequency;
}


ModIn& FunctionGenerator::getModIn(std::string modInName)
{
    if (modInName == "frequency")
        return frequency;
    else if (modInName == "phase")
        return phase;
    else if (modInName == "level")
        return level;
    else
        return frequency;
}

ModOut& FunctionGenerator::getModOut(std::string modOutName)
{
    if (modOutName == "mainOut")
        return mainOut;
    else if (modOutName == "auxOut1")
        return auxOut1;
    else if (modOutName == "auxOut2")
        return auxOut2;
    else if (modOutName == "auxOut3")
        return auxOut3;
    else
        return mainOut;
}
